finetrainers/trainer.py

```diff
@@ -0,0 +1,542 @@
+import inspect
+import json
+import logging
+import math
+import os
+import random
+import shutil
+from datetime import timedelta
+from typing import Any, Dict
+from pathlib import Path
+
+import diffusers
+import torch
+import torch.backends
+import transformers
+from accelerate import Accelerator, DistributedType
+from accelerate.logging import get_logger
+from accelerate.utils import DistributedDataParallelKwargs, InitProcessGroupKwargs, ProjectConfiguration, set_seed
+from diffusers.optimization import get_scheduler
+from diffusers.training_utils import cast_training_params, compute_density_for_timestep_sampling, compute_loss_weighting_for_sd3
+from huggingface_hub import create_repo, upload_folder
+from peft import LoraConfig, get_peft_model_state_dict, set_peft_model_state_dict
+from tqdm import tqdm
+
+from .args import Args
+from .constants import FINETRAINERS_LOG_LEVEL
+from .dataset import BucketSampler, VideoDatasetWithResizing
+from .models import get_config_from_model_name
+from .state import State
+from .utils.optimizer_utils import get_optimizer, gradient_norm
+from .utils.memory_utils import bytes_to_gigabytes, make_contiguous
+from .utils.torch_utils import unwrap_model
+
+
+logger = get_logger("finetrainers")
+logger.setLevel(FINETRAINERS_LOG_LEVEL)
+
+class Trainer:
+    def __init__(self, args: Args) -> None:
+        self.args = args
+        _validate_args(args)
+        
+        self.state = State()
+        
+        # Tokenizers
+        self.tokenizer = None
+        self.tokenizer_2 = None
+        self.tokenizer_3 = None
+        
+        # Text encoders
+        self.text_encoder = None
+        self.text_encoder_2 = None
+        self.text_encoder_3 = None
+        
+        # Denoisers
+        self.transformer = None
+        self.unet = None
+
+        # Autoencoders
+        self.vae = None
+
+        self._init_distributed()
+        self._init_logging()
+        self._init_directories_and_repositories()
+
+        self.state.model_name = self.args.model_name
+        self.model_config = get_config_from_model_name(self.args.model_name)
+    
+    def get_memory_statistics(self, precision: int = 3) -> Dict[str, Any]:
```

Comment:
I think this could be moved to `utils`. Or we could use class to keep track of it:
https://github.com/huggingface/peft/blob/ae55fdcc5c4830e0f9fb6e56f16555bafca392de/examples/oft_dreambooth/train_dreambooth.py#L421

Personally, I like the latter approach. Will also check with `accelerate` folks if they can ship it from `accelerate`.

Reply:
Yeah, sounds good. For now, I've moved it to a different file `memory_utils.py`. We can refactor later